streaming/src/main/scala/org/apache/spark/streaming/scheduler/ReceivedBlockTracker.scala

```diff
@@ -106,6 +106,12 @@ private[streaming] class ReceivedBlockTracker(
       timeToAllocatedBlocks(batchTime) = allocatedBlocks
       lastAllocatedBatchTime = batchTime
       allocatedBlocks
+    } else if (batchTime == lastAllocatedBatchTime) {
+      // This situation occurs when WAL is ended with BatchAllocationEvent,
+      // but without BatchCleanupEvent, possibly processed batch job or half-processed batch
+      // job need to process again, so the batchTime will be equal to lastAllocatedBatchTime.
+      // This situation will only occurs in recovery time.
+      logWarning(s"Possibly processed batch $batchTime need to be processed again in WAL recovery")
     } else {
```

Comment:
IIUC, I think this might not be happened normally, unless the checkpoint operation is so slow. Anyway we should consider this to make the code more robust.